analyzers/src/SonarAnalyzer.CSharp/Rules/GenericLoggerInjectionShouldMatchEnclosingType.cs

```diff
@@ -0,0 +1,57 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class GenericLoggerInjectionShouldMatchEnclosingType : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6672";
+    private const string MessageFormat = "Update this logger to use its enclosing type.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            var constructor = (ConstructorDeclarationSyntax)c.Node;
+            foreach (var invalidType in InvalidTypeParameters(constructor, c.SemanticModel))
+            {
+                c.ReportIssue(Diagnostic.Create(Rule, invalidType.GetLocation()));
+            }
+        },
+        SyntaxKind.ConstructorDeclaration);
+
+    // Returns T for [Constructor(ILogger<T> logger)] where T is not Constructor
+    private static IEnumerable<TypeSyntax> InvalidTypeParameters(ConstructorDeclarationSyntax constructor, SemanticModel model) =>
+        constructor.ParameterList.Parameters
+        .Where(x => x.Type is GenericNameSyntax)
+        .Select(x => (GenericNameSyntax)x.Type)
+        .Where(x => x.TypeArgumentList.Arguments.Count == 1
+                    && x.TypeArgumentList.Arguments[0].GetIdentifier()?.ValueText != constructor.Identifier.ValueText
```

Comment:
I cannot think of a reason to use semantics here.
The constructor _has_ to have the same name as the enclosing type, so if the generic is not exactly the same string as the constructor name, it's not using the correct type.

If there is a way to have a mismatch like this, please let me know:

```csharp

class Foo
{
    Bar(ILogger<Foo> logger) // Bar is a constructor (?)
    {}
}

Reply:
You have a test case close to this one, but I would still add a test case with a type that has a constructor that needs to be called by inheritors.